x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -263,18 +249,56 @@ public IndexPrivilege privilege() {
             return query;
         }
 
+        public FieldPermissions getFieldPermissions() {
+            return fieldPermissions;
+        }
+
         private boolean check(String action) {
             return actionMatcher.test(action);
         }
 
         private boolean check(String action, String index) {
             assert index != null;
-            return check(action) && indexNameMatcher.test(index);
+            return check(action) && (indexNameMatcher.test(index)
+                    && ((false == RestrictedIndicesNames.NAMES_SET.contains(index)) // if it is not restricted no more checks are required
+                            || IndexPrivilege.MONITOR.predicate().test(action) // allow monitor as a special case, even for restricted
+                            || allowRestrictedIndices)
+                    );
         }
 
         boolean hasQuery() {
             return query != null;
         }
+
+        public boolean allowRestrictedIndices() {
+            return allowRestrictedIndices;
+        }
+
+        public static Automaton buildIndexMatcherAutomaton(boolean allowRestrictedIndices, String... indices) {
+            final Automaton indicesAutomaton = Automatons.patterns(indices);
+            if (allowRestrictedIndices) {
+                return indicesAutomaton;
+            } else {
+                return Automatons.minusAndMinimize(indicesAutomaton, RestrictedIndicesNames.NAMES_AUTOMATON);
+            }
+        }
+
+        private static Predicate<String> buildIndexMatcherPredicateForAction(String action, Group... groups) {
+            final Set<String> ordinaryIndices = new HashSet<>();
+            final Set<String> restrictedIndices = new HashSet<>();
+            for (final Group group : groups) {
+                if (group.actionMatcher.test(action)) {
+                    if (group.allowRestrictedIndices) {
+                        restrictedIndices.addAll(Arrays.asList(group.indices()));
+                    } else {
+                        ordinaryIndices.addAll(Arrays.asList(group.indices()));
+                    }
+                }
+            }
+            return indexMatcher(restrictedIndices)
```

Comment:
> I think it might be worth checking if we have any restricted or any ordinary

Agreed will do!

> shouldn't we take into account the monitor action case?

Ha! I noticed this too, but I decided to keep the original behavior. In the original behavior, monitoring for wildcards will not work, indices have to be named explicitly.

Out of the three options:
1. remove support all together
2. keep it as is
3. extend monitoring priv to wildcards

I am leaning towards 1, but only slightly. Maintaining the status quo was chosen so that this PR sail smoothly.

I think we can make any such change in a different PR (if it's breaking I'll take good care of it to catch the train!)

Reply:
Let's maintain status quo in this PR and then do 1 in a followup targeted towards 7.0